lib/Onyx.d.ts

```diff
@@ -0,0 +1,258 @@
+ * safe for removal.
+ *
+ * @param testKey
+ */
+declare function isSafeEvictionKey<K extends Key>(testKey: K): boolean;
+
+/**
+ * Removes a key previously added to this list
+ * which will enable it to be deleted again.
+ *
+ * @param key
+ * @param connectionID
+ */
+declare function removeFromEvictionBlockList<K extends Key>(key: K, connectionID: number): void;
+
+/**
+ * Keys added to this list can never be deleted.
+ *
+ * @param key
+ * @param connectionID
+ */
+declare function addToEvictionBlockList<K extends Key>(key: K, connectionID: number): void;
+
+/**
+ * Subscribes a react component's state directly to a store key
+ *
+ * @example
+ * const connectionID = Onyx.connect({
+ *     key: ONYXKEYS.SESSION,
+ *     callback: onSessionChange,
+ * });
+ *
+ * @param mapping the mapping information to connect Onyx to the components state
+ * @param mapping.key ONYXKEY to subscribe to
+ * @param [mapping.statePropertyName] the name of the property in the state to connect the data to
+ * @param {Object} [mapping.withOnyxInstance] whose setState() method will be called with any changed data
+ *      This is used by React components to connect to Onyx
+ * @param [mapping.callback] a method that will be called with changed data
+ *      This is used by any non-React code to connect to Onyx
+ * @param [mapping.initWithStoredValues] If set to false, then no data will be prefilled into the
+ *  component
+ * @param [mapping.waitForCollectionCallback] If set to true, it will return the entire collection to the callback as a single object
+ * @param {String|Function} [mapping.selector] THIS PARAM IS ONLY USED WITH withOnyx(). If included, this will be used to subscribe to a subset of an Onyx key's data.
+ *       If the selector is a string, the selector is passed to lodashGet on the sourceData. If the selector is a function, the sourceData and withOnyx state are
+ *       passed to the selector and should return the simplified data. Using this setting on `withOnyx` can have very positive performance benefits because the component
+ *       will only re-render when the subset of data changes. Otherwise, any change of data on any property would normally cause the component to re-render (and that can
+ *       be expensive from a performance standpoint).
+ * @returns an ID to use when calling disconnect
+ */
+declare function connect<K extends Key>(mapping: ConnectOptions<K>): number;
+
+/**
+ * Remove the listener for a react component
+ * @example
+ * Onyx.disconnect(connectionID);
+ *
+ * @param connectionID unique id returned by call to Onyx.connect()
+ * @param [keyToRemoveFromEvictionBlocklist]
+ */
+declare function disconnect<K extends Key>(connectionID: number, keyToRemoveFromEvictionBlocklist?: K): void;
+
+/**
+ * Write a value to our store with the given key
+ *
+ * @param key ONYXKEY to set
+ * @param value value to store
+ */
+declare function set<K extends Key>(key: K, value: Value[K]): Promise<void>;
```

Comment:
Yes that is a question that I also have, I think the values are `null` when they key is "empty", when confirming this we should replace all `undefined` with `null` as well.

Reply:
@blazejkustra @fabioh8010 

Should we specify which key-value pairs can be optional in `OnyxValues`? Or are all keys optional?

And in the client code, do we pass `null` or `undefined` when we want to clear data for a key?

Does `set` do special operations if the `value` is `null` (for example, removing keys completely instead of just assigning `null` to an existing key)?

Reply:
> Should we specify which key-value pairs can be optional in OnyxValues? Or are all keys optional?

I'm leaning towards required key-value pairs with default values provided in `Onyx.init()`, so that we don't have to provide default values (empty object most of the time`{}`) in every component that uses `withOnyx` HOC. For now I'm typing some of OnyxValues with undefined but we can change that. Wdyt?

> And in the client code, do we pass null or undefined when we want to clear data for a key?

Probably both? @fabioh8010 let's check how it is implemented in onyx. We could ask Onyx engs what is the expected usage.

Reply:
@hayata-suenaga @blazejkustra 

> And in the client code, do we pass null or undefined when we want to clear data for a key?

To clear the key, you have to use `set` and pass `null` (we **can't** pass `undefined`).

```ts
Onyx.set(ONYXKEYS.ACCOUNT, null);
```

> Does set do special operations if the value is null (for example, removing keys completely instead of just assigning null to an existing key)?

It will remove the key from the storage.

> Should we specify which key-value pairs can be optional in OnyxValues? Or are all keys optional?
> 
> I'm leaning towards required key-value pairs with default values provided in Onyx.init(), so that we don't have to provide default values (empty object most of the time{}) in every component that uses withOnyx HOC. For now I'm typing some of OnyxValues as optional but we can change that. Wdyt?

We should type `OnyxValues` with all the root keys as required ones e.g.:

```ts
type Account = {
  /** URL to the assigned guide's appointment booking calendar */
  guideCalendarLink?: string;

  /** User recovery codes for setting up 2-FA */
  recoveryCodes?: string;

  /** Secret key to enable 2FA within the authenticator app */
  twoFactorAuthSecretKey?: string;
};

type OnyxValues = {
  [ONYXKEYS.ACCOUNT]: Account; // must be required here.
  [ONYXKEYS.ACCOUNT_MANAGER_REPORT_ID]: string;
  [ONYXKEYS.NVP_IS_FIRST_TIME_NEW_EXPENSIFY_USER]: boolean;
  [ONYXKEYS.ACTIVE_CLIENTS]: string[];
  [ONYXKEYS.DEVICE_ID]: string;
};
```

The reason for this is that the Onyx `connect` function and `withOnyx` HOC types will already return any value with the possibility of being `null`, because we can't guarantee that any key will always have a value set, even if it was defined a default one in `Onyx.init()` because I can simple set them to `null` later.

Just FYI that I have been studying the Onyx codebase and internals to understand better and ensure our typings are correct.

Reply:
@fabioh8010 thank you for all the research you're doing in Onyx. I have one more question

> Does set do special operations if the value is null (for example, removing keys completely instead of just assigning null to an existing key)?

>> It will remove the key from the storage.

So when you set a value to `null` for a key and later tries to retrieve value for the key, it's actually `undefined` because they key doesn't exist anymore?

@fabioh8010 and let me know if you need help understanding Onyx more. I can connect you to engineers who wrote the library

Reply:
@hayata-suenaga 

> So when you set a value to null for a key and later tries to retrieve value for the key, it's actually undefined because they key doesn't exist anymore?

So, this is some of the inconsistencies that I found while exploring the library (left a [message on Slack about some strange behaviors that I will report later](https://expensify.slack.com/archives/C0593T50UHG/p1688480256190489)). When using the `withOnyx` HOC and the key is empty, the value returned is `undefined`. When using the `connect` function, the value returned is `null`. I don't know if there is a reason for that but we can discuss when I post those findings in the open source channel.

> @fabioh8010 and let me know if you need help understanding Onyx more. I can connect you to engineers who wrote the library

Thank you for that @hayata-suenaga ! Sure, I will post any questions I have in the TS or open source channel in case I'm unable to understand and it's blocking us.